components/calendar/calendar.js

```diff
@@ -0,0 +1,480 @@
+
+	// populate first week of month
+	const firstWeek = [];
+	const numDaysFromPrevMonthToShow = getNumberOfDaysFromPrevMonthToShow(shownMonth, shownYear);
+
+	if (numDaysFromPrevMonthToShow > 0) {
+		const prevMonth = getPrevMonth(shownMonth);
+		const prevMonthYear = (shownMonth === 0) ? (shownYear - 1) : shownYear;
+		const numDaysLastMonth = getNumberOfDaysInMonth(prevMonth, prevMonthYear);
+		for (let i = numDaysLastMonth - numDaysFromPrevMonthToShow + 1; i <= numDaysLastMonth; i++) {
+			firstWeek.push(new Date(prevMonthYear, prevMonth, i));
+		}
+	}
+	for (let j = 1; j <= daysInWeek - numDaysFromPrevMonthToShow; j++) {
+		firstWeek.push(new Date(shownYear, shownMonth, j));
+	}
+	dates.push(firstWeek);
+
+	// remaining weeks
+	let nextMonthDay = 1;
+	let firstDateOfWeek = daysInWeek - numDaysFromPrevMonthToShow + 1;
+	const numWeeks = Math.ceil((numDaysFromPrevMonthToShow + numDays) / daysInWeek);
+	const nextMonth =  getNextMonth(shownMonth);
+	const nextMonthYear = (shownMonth === 11) ? (shownYear + 1) : shownYear;
+	for (let i = 1; i < numWeeks; i++) {
+		const week = [];
+		for (let j = firstDateOfWeek; j < firstDateOfWeek + daysInWeek; j++) {
+			let day;
+			if (j > numDays) {
+				day = new Date(nextMonthYear, nextMonth, nextMonthDay);
+				nextMonthDay++;
+			} else {
+				day = new Date(shownYear, shownMonth, j);
+			}
+			week.push(day);
+		}
+		firstDateOfWeek = firstDateOfWeek + daysInWeek;
+		dates.push(week);
+	}
+
+	return dates;
+}
+
+export function getNextMonth(month) {
+	return (month === 11) ? 0 : (month + 1);
+}
+
+export function getNumberOfDaysFromPrevMonthToShow(month, year) {
+	const firstDayOfMonth = new Date(year, month, 1).getDay();
+	let numDaysFromLastMonthToShowThisMonth = 0;
+	if (firstDayOfMonth !== firstDayOfWeek) {
+		numDaysFromLastMonthToShowThisMonth = firstDayOfMonth - firstDayOfWeek;
+		if (numDaysFromLastMonthToShowThisMonth < 0) {
+			numDaysFromLastMonthToShowThisMonth += daysInWeek;
+		}
+	}
+	return numDaysFromLastMonthToShowThisMonth;
+}
+
+export function getNumberOfDaysInMonth(month, year) {
+	return new Date(year, month + 1, 0).getDate();
+}
+
+export function getNumberOfDaysToSameWeekPrevMonth(month, year) {
+	const prevMonth = getPrevMonth(month);
+	const prevMonthYear = prevMonth === 11 ? year - 1 : year;
+	const numDaysPrevMonth = getNumberOfDaysInMonth(prevMonth, prevMonthYear);
+	const numDaysFromPrevPrevMonthToShowPrevMonth = getNumberOfDaysFromPrevMonthToShow(prevMonth, prevMonthYear);
+	const numWeeksPrevMonth = Math.ceil((numDaysFromPrevPrevMonthToShowPrevMonth + numDaysPrevMonth) / 7);
+
+	const firstDayOfMonth = new Date(year, month, 1).getDay();
+	const hasDaysFromPrevMonth = (firstDayOfMonth !== firstDayOfWeek);
+	return daysInWeek * (numWeeksPrevMonth - ((hasDaysFromPrevMonth) ? 1 : 0));
+}
+
+export function getPrevMonth(month) {
+	return (month === 0) ? 11 : (month - 1);
+}
+
+export function parseDate(val) {
+	if (!val) return null;
+
```

Comment:
Wondering if maybe a regex might make this a bit easier to validate and then extract the various bits using groups? That's what the time input is doing and was pretty straightforward.

Naive attempt:
`/^([0-9]{4})-([0-9]{2})-([0-9]{2})/`

That way it'll match whether the time is there or not and you can use the groups + `parseInt` to grab the year, month and day.

Reply:
We could also use [Date.parse()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse)... but it supports all kinds of crazy formats including those with timezone offsets which we maybe don't want if we want to force this into UTC.